### PR TITLE
feat: allow adding additional members to production parse methods

### PR DESCRIPTION
**Pull request by pyoor**
This PR adds the `extMembers` key to the production parse method's option bag.  The `extMembers` key is used to add additional members that may exist during parsing.  This is especially useful for allowing custom productions expand existing productions.

Additionally, this PR adds `extensions` to the parse method so that users can add additional allowedMembers on a per production basis.

### CODE REVIEW

File: lib/webidl2.js [hunk 1]
```diff
@@ -32,28 +44,42 @@ function parseByTokens(tokeniser, options) {
     const callback = consume("callback");
     if (!callback) return;
     if (tokeniser.probe("interface")) {
-      return CallbackInterface.parse(tokeniser, callback);
+      return CallbackInterface.parse(tokeniser, callback, {
+        ...options?.extensions?.["callback-interface"],
```
Comment (saschanaz) [inline lib/webidl2.js:48]: callbackInterface maybe?

File: lib/webidl2.js [hunk 2]
```diff
@@ -72,11 +98,11 @@ function parseByTokens(tokeniser, options) {
       callback() ||
       interface_() ||
       partial() ||
-      Dictionary.parse(tokeniser) ||
+      Dictionary.parse(tokeniser, { ...options?.extensions?.dictionary }) ||
```
Comment (saschanaz) [inline lib/webidl2.js:101]:
```suggestion
      Dictionary.parse(tokeniser, options?.extensions?.dictionary) ||
```

Probably no need to spread at all. Same for namespace below.

File: test/extensions.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import expect from "expect";
+import { Attribute } from "../lib/productions/attribute.js";
+
+const extensions = {
+  "callback-interface": {
+    extMembers: [[Attribute.parse]],
+  },
+  namespace: {
+    extMembers: [[Attribute.parse, { readonly: false }]],
+  },
```
Comment (saschanaz) [inline test/extensions.js:11]: I think it's better to have unit tests, like https://github.com/w3c/webidl2.js/blob/main/test/custom-production.js.

**Review by saschanaz — CHANGES_REQUESTED** [review]

File: test/extensions.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
+"use strict";
```
Comment (saschanaz) [inline test/extensions.js:1]: We don't need `"use strict"` here, it's the default in ES modules!

Comment (saschanaz) [reply test/extensions.js:1]: Oh then I found the test I gave you still has it. Anyway it's not needed 😛

Comment (pyoor) [reply test/extensions.js]: Removed in commit [f12005e](https://github.com/w3c/webidl2.js/pull/731/commits/f12005ee66c058dd155c628d63168043b04de1b1)

File: test/custom-production.js [hunk 5]
```diff
@@ -1,7 +1,40 @@
...
 import expect from "expect";
+
 import { parse, write } from "webidl2";
+import { Base } from "../lib/productions/base.js";
+import {
+  autoParenter,
+  type_with_extended_attributes,
+} from "../lib/productions/helpers.js";
+
+class CustomAttribute extends Base {
```
Comment (saschanaz) [inline test/custom-production.js:10]: Good, can you replace the existing test below to use this? I think this one is better than `const customProduction` in it.

File: test/util/collect.js [hunk 6]
```diff
@@ -7,7 +7,7 @@ import { diff } from "jsondiffpatch";
  * Collects test items from the specified directory
  * @param {string} base
  */
-export function* collect(base, { expectError, raw } = {}) {
+export function* collect(base, { expectError, raw, extensions = {} } = {}) {
```
Comment (saschanaz) [inline test/util/collect.js:10]: (I guess this is not needed anymore)

Comment (pyoor) [reply test/util/collect.js]: Addressed in [1b29589](https://github.com/w3c/webidl2.js/pull/731/commits/1b2958989a9fd81796d82b668b112222ef0d1648).

File: test/custom-production.js (hunk 5 above)
Comment (pyoor) [reply test/custom-production.js:10]: Addressed in [9cfc34c](https://github.com/w3c/webidl2.js/pull/731/commits/9cfc34c560c39960f13eb06b16c183e01bfd2165).

**Comment by saschanaz** [comment]
Oh, oops. Could you open this again targeting `main` branch? 👀

**Pull request closed without merging**